SYCL/SpecConstants/2020/handler-api.cpp

```diff
@@ -0,0 +1,199 @@
+// FIXME: SYCL 2020 specialization constants are not supported on host device
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// FIXME: ACC devices use emulation path, which is not yet supported
+// FIXME: CUDA uses emulation path, which is not yet supported
+// UNSUPPORTED: cuda
+
+#include <sycl/sycl.hpp>
+#include <iostream>
+#include <cstdlib>
+
+#include "common.hpp"
+
+constexpr sycl::specialization_id<int> int_id;
+constexpr sycl::specialization_id<double> double_id(3.14);
+constexpr sycl::specialization_id<custom_type> custom_type_id;
+
+class TestDefaultValuesKernel;
+class EmptyKernel;
+class TestSetAndGetOnDevice;
+
+bool test_default_values(sycl::queue q);
+bool test_set_and_get_on_host(sycl::queue q);
+bool test_set_and_get_on_device(sycl::queue q);
+
+int main() {
+  auto exception_handler = [&](sycl::exception_list exceptions) {
+    for (std::exception_ptr const &e : exceptions) {
+      try {
+        std::rethrow_exception(e);
+      } catch (sycl::exception const &e) {
+        std::cout << "An async SYCL exception was caught: " << e.what()
+                  << std::endl;
+        std::exit(1);
+      }
+    }
+  };
+
+  sycl::queue q(exception_handler);
+
+  if (!test_default_values(q)) {
+    std::cout << "Test for default values of specialization constants failed!"
+              << std::endl;
+    return 1;
+  }
+
+  if (!test_set_and_get_on_host(q)) {
+    std::cout << "Test for set and get API on host failed!"
+              << std::endl;
+    return 1;
+  }
+
+  if (!test_set_and_get_on_device(q)) {
+    std::cout << "Test for set and get API on device failed!"
+              << std::endl;
+    return 1;
+  }
+
+  return 0;
+};
+
+bool test_default_values(sycl::queue q) {
+  sycl::buffer<int> int_buffer(1);
+  sycl::buffer<double> double_buffer(1);
+  sycl::buffer<custom_type> custom_type_buffer(1);
+
+  q.submit([&](sycl::handler &cgh) {
+    auto int_acc = int_buffer.get_access<sycl::access::mode::write>(cgh);
+    auto double_acc = double_buffer.get_access<sycl::access::mode::write>(cgh);
+    auto custom_type_acc =
+        custom_type_buffer.get_access<sycl::access::mode::write>(cgh);
+    cgh.single_task<TestDefaultValuesKernel>([=](sycl::kernel_handler kh) {
+      int_acc[0] = kh.get_specialization_constant<int_id>();
+      double_acc[0] = kh.get_specialization_constant<double_id>();
+      custom_type_acc[0] = kh.get_specialization_constant<custom_type_id>();
+    });
+  });
+
+  auto int_acc = int_buffer.get_access<sycl::access::mode::read>();
+  if (!check_value(0, int_acc[0],
+             "integer specialization constant (defined without default value)"))
+    return false;
+
+  auto double_acc = double_buffer.get_access<sycl::access::mode::read>();
+  if (!check_value(3.14, double_acc[0], "double specialization constant"))
+    return false;
+
+  auto custom_type_acc =
+      custom_type_buffer.get_access<sycl::access::mode::read>();
+  const custom_type custom_type_ref;
```

Comment:
Nit: the default value of the `custom_type_id` coincides with default CPP (zeroes). the test could be strengthened by assigning custom_type_id some specific non-zero value to verify new SYCL2020 capabilities work for non-primitive types as well.

Reply:
> Nit: the default value of the `custom_type_id` coincides with default CPP (zeroes). 

It should have some non-zero default values, as I have its fields initialized: https://github.com/intel/llvm-test-suite/pull/255/files#diff-308baade624f8087060f981e03c24cffa77255b25f54deb6f7b7f1f6349a84afR16